cstore_reader.c

```diff
@@ -200,6 +214,85 @@ CStoreReadFooter(StringInfo tableFooterFilename)
 }
 
 
+/* DeserializeBlockData  deserializes requested data block for all columns. It
+ * uncompresses serialized data if necessary, sets deserialized buffer index at
+ * read state. The function also deallocates previous data buffer used to keep temporary
+ * uncompressed data buffer.
+ */
+static void
+DeserializeBlockData(TupleDesc tupleDescriptor, StripeData *stripeData,
+					 MemoryContext stripeReadContext, uint64 blockIndex,
+					 BlockDataBuffers **blockDataBuffersArray)
+{
+	int columnIndex = 0;
+	MemoryContext oldContext = MemoryContextSwitchTo(stripeReadContext);
+
+	for (columnIndex=0; columnIndex < stripeData->columnCount; columnIndex++)
+	{
+		ColumnData *columnData = stripeData->columnDataArray[columnIndex];
+
+		if (columnData != NULL)
+		{
+			ColumnBlockData *blockData = columnData->blockDataArray[blockIndex];
+			Form_pg_attribute *attributeFormArray = tupleDescriptor->attrs;
+			bool typeByValue = attributeFormArray[columnIndex]->attbyval;
+			int typeLength = attributeFormArray[columnIndex]->attlen;
+			char typeAlign = attributeFormArray[columnIndex]->attalign;
+			BlockDataBuffers * blockDataBuffers = blockDataBuffersArray[columnIndex];
+			StringInfo valueBuffer = NULL;
+
+			if (blockDataBuffers->serializedValueBuffer != NULL)
+			{
+				pfree(blockDataBuffers->serializedValueBuffer->data);
+				pfree(blockDataBuffers->serializedValueBuffer);
+				blockDataBuffers->serializedValueBuffer = NULL;
+			}
+
+			valueBuffer = DecompressBuffer(blockData->valueBuffer,
+										   blockData->valueCompressionType);
+
+			DeserializeBoolArray(blockData->existBuffer,
+								 blockDataBuffers->existsArray,
+								 blockData->rowCount);
+			DeserializeDatumArray(valueBuffer, blockDataBuffers->existsArray,
+								  blockData->rowCount, typeByValue, typeLength,
+								  typeAlign, blockDataBuffers->valueArray);
+
+			/* store newly allocated uncompressed data buffer to be freed later */
+			if (valueBuffer != blockData->valueBuffer)
+			{
+				blockDataBuffers->serializedValueBuffer = valueBuffer;
+			}
```

Comment:
1. agreed. we should also make it more explicit by calling DecompressBuffer() only if compression type is something other than COMPRESSION_NONE
2.  Since DecompressBuffer allocates memory for each decompression operation within stripe read memory context, buffers used for decompressing may pile up until memory context reset at next stripe read. This way we could reclaim memory during block changes. Non-compressed case is a bit different. Memory is already read into serialized data buffers (ColumnBlockBuffers->valueBuffer), Deserialized version (ColumnBlockData->valueArray) may contain pointers to serialized version. since it was already in memory, there is no additional memory is allocated and there is no need to free it right away.

3.  I did not understand the comment here. Could you explain a bit more ?